velox/functions/lib/Re2Functions.cpp

```diff
@@ -970,51 +969,229 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> re2ExtractSignatures() {
+      charKind_ = CharKind::kWildcard;
+    } else {
+      charKind_ = CharKind::kNormal;
+    }
+  }
+
+  /// Current index of the cursor.
+  char currentIndex() const {
+    return currentIndex_;
+  }
+
+  /// Char at current cursor.
+  char current() const {
+    return pattern_.data()[currentIndex_];
+  }
+
+  CharKind state() const {
+    return charKind_;
+  }
+
+  CharKind previousState() const {
+    return previousCharKind_;
+  }
+
+ private:
+  const StringView pattern_;
+  const std::optional<char> escapeChar_;
+
+  int32_t currentIndex_ = -1;
+  // State of current char.
+  CharKind charKind_ = CharKind::kNormal;
+  // State of previous char.
+  CharKind previousCharKind_ = CharKind::kNormal;
+};
+
+PatternMetadata determinePatternKind(
+    StringView pattern,
+    std::optional<char> escapeChar) {
+  int32_t patternLength = pattern.size();
+
+  // Index of the first % or _ character(not escaped).
+  int32_t wildcardStart = -1;
   // Index of the first character that is not % and not _.
-  vector_size_t fixedPatternStart = -1;
+  int32_t fixedPatternStart = -1;
   // Index of the last character in the fixed pattern, used to retrieve the
   // fixed string for patterns of type kSubstring.
-  vector_size_t fixedPatternEnd = 0;
+  int32_t fixedPatternEnd = -1;
+  // Count of wildcard character sequences in pattern.
+  size_t numWildcardSequences = 0;
   // Total number of % characters.
-  vector_size_t anyCharacterWildcardCount = 0;
+  size_t anyCharacterWildcardCount = 0;
   // Total number of _ characters.
-  vector_size_t singleCharacterWildcardCount = 0;
-  auto patternStr = pattern.data();
+  size_t singleCharacterWildcardCount = 0;
+
+  PatternStringIterator iterator{pattern, escapeChar};
+  bool fallbackToGeneric = false;
+
+  // Return false if failed to handle the new fixed pattern(need to fallback),
+  // true otherwise.
+  auto handleNewFixedPattern = [&]() -> bool {
+    // Record the first fixed pattern start.
+    if (fixedPatternStart == -1) {
+      fixedPatternStart = iterator.currentIndex();
+    } else {
+      // This is not the first fixed pattern, not supported, so fallback.
+      if (iterator.previousState() ==
+          PatternStringIterator::CharKind::kWildcard) {
+        fallbackToGeneric = true;
+        return false;
+      }
+    }
+
+    return true;
+  };
 
-  while (i < patternLength) {
-    if (patternStr[i] == '%' || patternStr[i] == '_') {
-      if (wildcardStart == -1) {
-        wildcardStart = i;
+  // Iterator through the pattern string to collect the stats for the simple
+  // patterns that we can optimize: e.g. the
+  while (iterator.hasNext()) {
+    iterator.next();
+    auto current = iterator.current();
+
+    if (iterator.previousState() == PatternStringIterator::CharKind::kEscape) {
+      // The char follows escapeChar can only be one of (%, _, escapeChar).
+      if (current != '%' && current != '_' && current != escapeChar) {
+        fallbackToGeneric = true;
```

Comment:
If we reach here, does it mean that the pattern is not valid? If so, should we fail?

Reply:
Yes, but seems we can not throw exception here, exception need to be set to the context? which we don't have access to here, so here we will fallback to generic and let generic implementation handle the exception.

```
auto error = std::make_exception_ptr(std::invalid_argument(
    "Escape character must be followed by '%', '_' or the escape character itself"));
context.setErrors(rows, error);
```